File: test/reglas_cuadrante_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

import 'package:inces_lms_app/core/reglas_cuadrante.dart';
import 'package:inces_lms_app/models/cuadrante.dart';

/// Pruebas de las reglas del cuadrante.
///
/// El segundo grupo es el que importa de verdad, y no es el de las funciones
/// puras: **contrasta las constantes de Dart con la migración de M3**. Sin él,
/// este archivo sólo probaría que el espejo coincide consigo mismo. Con él, si
/// alguien mueve la frontera de turnos en la base o renombra un día, la prueba
/// falla aquí en vez de que la rejilla pinte un bloque en la franja equivocada.
///
/// Es la misma técnica que `backend/test/reglas-cuadrante.test.ts` —leer la
/// migración como texto— aplicada al lado de Dart, porque son **dos** espejos
/// distintos y cada uno necesita su propia vigilancia.
void main() {
  group('turnos y bloques', () {
    test('el turno de un bloque sigue la frontera, para los doce bloques', () {
      for (var bloque = 1; bloque <= bloquesDeManana; bloque++) {
        expect(
          turnoDeBloque(bloque),
          Turno.manana,
          reason: 'el bloque $bloque es de la mañana',
        );
      }
      for (var bloque = bloquesDeManana + 1;
          bloque <= bloqueMaximo;
          bloque++) {
        expect(
          turnoDeBloque(bloque),
          Turno.tarde,
          reason: 'el bloque $bloque es de la tarde',
        );
      }
    });

    test('los bloques de los dos turnos cubren el día sin huecos ni solapes', () {
      final todos = [
        ...bloquesDelTurno(Turno.manana),
        ...bloquesDelTurno(Turno.tarde),
      ];

      // Cada bloque aparece exactamente una vez, y entre los dos turnos cubren
      // del 1 al 12. Un bloque en los dos turnos lo pintaría dos veces en la
      // rejilla; uno en ninguno desaparecería sin que nadie lo note.
      expect(todos, [for (var b = 1; b <= bloqueMaximo; b++) b]);
    });

    test('un turno desconocido en el JSON se deriva del bloque', () {
      // La columna es `generated always as (turno_de_bloque(block)) stored`, así
      // que siempre llega. Si llegara corrupta, inventar «mañana» para un bloque
      // 9 pintaría la clase en la franja equivocada (R-16).
      expect(Turno.desdeApi(null, bloque: 3), Turno.manana);
      expect(Turno.desdeApi(null, bloque: 9), Turno.tarde);
      expect(Turno.desdeApi('', bloque: 9), Turno.tarde);
      expect(Turno.desdeApi('TARDE', bloque: 1), Turno.tarde);
      expect(Turno.desdeApi('MAÑANA', bloque: 12), Turno.manana);
    });
  });

  group('los espejos, contra la migración real', () {
    // Los valores que se comprueban, leídos del SQL y no escritos aquí.
    late int corteEnSql;
    late String mananaEnSql;
    late String tardeEnSql;
    late Map<int, String> diasEnSql;
    late List<int> diasMaximosEnSql;
    late List<int> bloquesMaximosEnSql;

    setUpAll(() {
      final sql = _leerMigracion();

      // `turno_de_bloque`: la frontera y los dos nombres de turno.
      final turno = _cuerpoDeFuncion(sql, 'turno_de_bloque');
      final corte = RegExp(
        r"p_bloque\s*<=\s*(\d+)\s+then\s+'([^']+)'\s+else\s+'([^']+)'",
      ).firstMatch(turno);

      if (corte == null) {
        fail('No se encontró el corte de `turno_de_bloque` en la migración.');
      }

      corteEnSql = int.parse(corte.group(1)!);
      mananaEnSql = corte.group(2)!;
      tardeEnSql = corte.group(3)!;

      // `dia_legible`: los seis nombres.
      final dia = _cuerpoDeFuncion(sql, 'dia_legible');
      diasEnSql = {
        for (final m in RegExp(r"when (\d) then '([^']+)'").allMatches(dia))
          int.parse(m.group(1)!): m.group(2)!,
      };

      // Los `check` de rango, que aparecen una vez por tabla (`teacher_duties`
      // y `schedule_slots`).
      diasMaximosEnSql = [
        for (final m in RegExp(r'check \(day_of_week between 1 and (\d+)\)')
            .allMatches(sql))
          int.parse(m.group(1)!),
      ];
      bloquesMaximosEnSql = [
        for (final m in RegExp(r'check \(block between 1 and (\d+)\)')
            .allMatches(sql))
          int.parse(m.group(1)!),
      ];
    });

    test('el analizador encontró lo que buscaba', () {
      // Suelo explícito. Sin él, un cambio de formato en la migración dejaría
      // los conjuntos vacíos y las comprobaciones de abajo pasarían por la
      // puerta de atrás.
      expect(diasEnSql.length, 6, reason: 'los seis días de `dia_legible`');
      expect(
        diasMaximosEnSql.length,
        greaterThanOrEqualTo(2),
        reason: 'el `check` de día está en las dos tablas del cuadrante',
      );
      expect(
        bloquesMaximosEnSql.length,
        greaterThanOrEqualTo(2),
        reason: 'el `check` de bloque está en las dos tablas del cuadrante',
      );
    });

    test('la frontera de turnos es la misma que la de la base', () {
      expect(
        bloquesDeManana,
        corteEnSql,
        reason: 'la frontera de turnos se movió en la base y no en Dart',
      );
      expect(turnoDeBloque(1).valorApi, mananaEnSql);
      expect(turnoDeBloque(bloqueMaximo).valorApi, tardeEnSql);
    });

    test('los días son los mismos que los de la base, y en el mismo orden', () {
      // Se compara sin distinguir mayúsculas a propósito: el SQL escribe
      // «lunes» porque va dentro de una frase («Ese docente ya tiene… el
      // lunes»), y la interfaz escribe «Lunes» porque encabeza una columna.
      // Lo que no puede diferir es el nombre ni el orden.
      for (var dia = 1; dia <= diaMaximo; dia++) {
        expect(
          diasDeLaSemana[dia].toLowerCase(),
          diasEnSql[dia],
          reason: 'el día $dia no coincide con el de la base',
        );
      }
    });

    test('los máximos de día y bloque son los de la base', () {
      for (final maximo in diasMaximosEnSql) {
        expect(diaMaximo, maximo, reason: 'el máximo de día no coincide');
      }
      for (final maximo in bloquesMaximosEnSql) {
        expect(bloqueMaximo, maximo, reason: 'el máximo de bloque no coincide');
      }
    });

    test('las dos tablas del cuadrante usan el mismo rango', () {
      // Si una tabla admitiera el día 7 y la otra no, el mismo hueco sería
      // válido en una y rechazado en la otra, y el chequeo de colisiones
      // compararía rangos distintos.
      expect(diasMaximosEnSql.toSet(), hasLength(1));
      expect(bloquesMaximosEnSql.toSet(), hasLength(1));
    });
  });

  group('fechas', () {
    test('acepta una fecha ISO bien formada', () {
      expect(esFechaISO('2026-09-21'), isTrue);
      expect(esFechaISO('2027-02-28'), isTrue);
      expect(esFechaISO('2028-02-29'), isTrue, reason: '2028 es bisiesto');
      expect(esFechaISO('  2026-09-21  '), isTrue, reason: 'se recorta');
    });

    test('rechaza un día que no existe, no sólo un formato raro', () {
      // Éste es el caso que una comprobación de formato dejaría pasar.
      expect(esFechaISO('2027-02-30'), isFalse);
      expect(esFechaISO('2027-04-31'), isFalse);
      expect(esFechaISO('2027-02-29'), isFalse, reason: '2027 no es bisiesto');
    });

    test('rechaza formatos que no son AAAA-MM-DD', () {
      expect(esFechaISO('21/09/2026'), isFalse);
      expect(esFechaISO('2026-9-21'), isFalse);
      expect(esFechaISO('2026-09-21T00:00:00Z'), isFalse);
      expect(esFechaISO(''), isFalse);
    });

    test('un rango a medias es válido, porque las fechas son opcionales', () {
      // El centro no ha cargado las fechas del lapso en curso e inventarlas
      // sería fabricar un dato institucional (R-17).
      expect(rangoDeFechasValido(inicio: null, fin: null), isTrue);
      expect(rangoDeFechasValido(inicio: '2026-09-21', fin: null), isTrue);
      expect(rangoDeFechasValido(inicio: null, fin: '2027-02-13'), isTrue);
    });

    test('exige que el fin sea posterior al inicio', () {
      expect(
        rangoDeFechasValido(inicio: '2026-09-21', fin: '2027-02-13'),
        isTrue,
      );
      expect(
        rangoDeFechasValido(inicio: '2026-09-21', fin: '2026-09-21'),
        isFalse,
        reason: 'un lapso de un día no es un rango',
      );
      expect(
        rangoDeFechasValido(inicio: '2027-02-13', fin: '2026-09-21'),
        isFalse,
      );
    });
  });

  group('códigos, nombres y capacidades', () {
    test('acepta los códigos de lapso que usa el centro', () {
      expect(esCodigoDeLapso('2026-1'), isTrue);
      expect(esCodigoDeLapso('SA26-2'), isTrue);
      expect(esCodigoDeLapso('2026-2'), isTrue);
    });

    test('rechaza un código de lapso mal formado', () {
      expect(esCodigoDeLapso(''), isFalse);
      expect(esCodigoDeLapso('-2026'), isFalse, reason: 'empieza por guion');
      expect(esCodigoDeLapso('2026 1'), isFalse, reason: 'lleva un espacio');
      expect(esCodigoDeLapso('2026_1'), isFalse);
      expect(esCodigoDeLapso('2026-1234567'), isFalse, reason: 'pasa de 10');
    });

    test('el nombre de un espacio no puede estar en blanco', () {
      expect(nombreDeAulaValido('Taller de Soldadura'), isTrue);
      expect(nombreDeAulaValido(''), isFalse);
      expect(nombreDeAulaValido('   '), isFalse);
      expect(nombreDeAulaValido('a' * 80), isTrue);
      expect(nombreDeAulaValido('a' * 81), isFalse);
    });

    test('la capacidad cero es válida y la negativa no', () {
      // Cero significa «sin cupo declarado»: una zona, un pasillo. No es lo
      // mismo que desconocido, y por eso la columna es `not null default 0`.
      expect(capacidadValida(0), isTrue);
      expect(capacidadValida(12), isTrue);
      expect(capacidadValida(-1), isFalse);
    });

    test('el tipo de un espacio sale de sus dos columnas', () {
      expect(TipoAula.deAula(esTaller: true, capacidad: 12), TipoAula.taller);
      expect(TipoAula.deAula(esTaller: true, capacidad: 0), TipoAula.taller);
      expect(TipoAula.deAula(esTaller: false, capacidad: 12), TipoAula.aula);
      expect(TipoAula.deAula(esTaller: false, capacidad: 0), TipoAula.zona);
    });

    test('los tres tipos son excluyentes y cubren todos los casos', () {
      for (final esTaller in [true, false]) {
        for (final capacidad in [0, 1, 30]) {
          final tipo = TipoAula.deAula(
            esTaller: esTaller,
            capacidad: capacidad,
          );
          // El filtro de la pantalla se elige entre estos tres, así que ninguno
          // puede quedar fuera sin que el usuario lo sepa.
          expect(TipoAula.values, contains(tipo));
        }
      }
    });
  });

  group('días legibles', () {
    test('nombra los seis días del centro', () {
      expect(diaLegible(1), 'Lunes');
      expect(diaLegible(6), 'Sábado');
    });

    test('un día fuera de rango no se inventa un nombre', () {
      // «Día 9» se leería como un dato raro que enseñar; un `null` deja que la
      // pantalla lo trate como el error que es.
      expect(diaLegible(0), isNull);
      expect(diaLegible(7), isNull);
      expect(diaLegible(-1), isNull);
    });
  });

  group('choque de agenda', () {
    test('se reconoce por el código, no por el tipo de error', () {
      // `ApiClient` clasifica todo 409 como `AppErrorType.validacion`, así que
      // el tipo no distingue un choque de un dato mal formado. Es el código el
      // que permite dar un mensaje útil.
      expect(esChoqueDeAgenda(codigoChoqueDeAgenda), isTrue);
      expect(esChoqueDeAgenda('RESTRICCION_VIOLADA'), isFalse);
      expect(esChoqueDeAgenda('REGISTRO_DUPLICADO'), isFalse);
      expect(esChoqueDeAgenda(null), isFalse);
    });
  });

  group('el nombre de un docente sin nombre', () {
    test('se nombra el hueco en vez de dejarlo mudo', () {
      // R-21: el canal de invitación nunca captura los nombres, así que la vista
      // devuelve NULL. Un hueco en blanco se lee como un fallo de la aplicación.
      expect(nombreDeDocente('Luis Márquez'), 'Luis Márquez');
      expect(nombreDeDocente(''), 'Docente sin nombre');
      expect(nombreDeDocente('   '), 'Docente sin nombre');
      expect(nombreDeDocente(null), 'Docente sin nombre');
    });
  });
}

// -----------------------------------------------------------------------------
//  Lectura de la migración
// -----------------------------------------------------------------------------

const String _rutaMigracion =
    'supabase/migrations/202609180001_mod3_cuadrante_aulas.sql';

/// Localiza la migración de M3 sin dar por hecho el directorio de trabajo.
///
/// `flutter test` corre desde la raíz del paquete, pero asumirlo convertiría un
/// cambio de invocación en un «fichero no encontrado» que no explica nada.
File _archivoMigracion() {
  var directorio = Directory.current;

  for (var intentos = 0; intentos < 6; intentos++) {
    final candidato = File('${directorio.path}/$_rutaMigracion');
    if (candidato.existsSync()) return candidato;

    final padre = directorio.parent;
    if (padre.path == directorio.path) break;
    directorio = padre;
  }

  fail(
    'No se encontró $_rutaMigracion subiendo desde ${Directory.current.path}',
  );
}

String _leerMigracion() => _archivoMigracion().readAsStringSync();

/// Extrae el cuerpo de una función `create or replace function public.X`.
String _cuerpoDeFuncion(String sql, String nombre) {
  final patron = RegExp(
    'create or replace function public\\.$nombre\\b[\\s\\S]*?\\\$\\\$'
    '([\\s\\S]*?)\\\$\\\$',
    caseSensitive: false,
  );
  final coincidencia = patron.firstMatch(sql);

  if (coincidencia == null) {
    fail('No se encontró la función $nombre en la migración de M3.');
  }

  return coincidencia.group(1) ?? '';
}

File: backend/test/reglas-cuadrante.test.ts
import { describe, expect, it } from 'vitest';
import {
  BLOQUE_MAXIMO,
  DIA_MAXIMO,
  diaLegible,
  esChoqueDeAgenda,
  esFechaISO,
  rangoDeFechasValido,
  turnoDeBloque,
} from '../src/dominio/reglas-cuadrante.js';

/**
 * Las reglas puras de M3, sin montar HTTP ni repositorios.
 *
 * Se prueban aquí por la misma razón que `reglas-curriculo.test.ts`: una regla
 * enterrada en un manejador sólo se ejercita levantando la API entera. Como
 * función pura se prueban los casos que la API casi nunca alcanza —el día 7, el
 * bloque 13, el 30 de febrero— que son justo los que aparecen el día que alguien
 * pega una planilla desde una hoja de cálculo.
 */

describe('turnoDeBloque', () => {
  it('los seis primeros bloques son la mañana', () => {
    for (let bloque = 1; bloque <= 6; bloque += 1) {
      expect(turnoDeBloque(bloque), `bloque ${bloque}`).toBe('MAÑANA');
    }
  });

  it('del séptimo en adelante es la tarde', () => {
    for (let bloque = 7; bloque <= BLOQUE_MAXIMO; bloque += 1) {
      expect(turnoDeBloque(bloque), `bloque ${bloque}`).toBe('TARDE');
    }
  });

  it('el corte está exactamente en el 6/7, no antes ni después', () => {
    // Esta aserción es la que avisa si alguien mueve la frontera en la función
    // SQL y olvida este espejo. El corte es provisional (R-16), así que se
    // moverá; lo que no puede pasar es que los dos sitios se separen en
    // silencio y el backend calcule un turno que la base nunca produciría.
    expect(turnoDeBloque(6)).toBe('MAÑANA');
    expect(turnoDeBloque(7)).toBe('TARDE');
  });
});

describe('diaLegible', () => {
  it('nombra los seis días laborables', () => {
    expect(diaLegible(1)).toBe('lunes');
    expect(diaLegible(2)).toBe('martes');
    expect(diaLegible(3)).toBe('miércoles');
    expect(diaLegible(4)).toBe('jueves');
    expect(diaLegible(5)).toBe('viernes');
    expect(diaLegible(6)).toBe('sábado');
  });

  it('el domingo NO tiene nombre: el requisito es de lunes a sábado', () => {
    // 7 es domingo. Devolver «domingo» aquí sugeriría que el día es válido,
    // cuando el `check (day_of_week between 1 and 6)` lo rechaza.
    expect(diaLegible(7)).toBe('día 7');
  });

  it('un día imposible devuelve texto en vez de `undefined`', () => {
    // El `check` de la base ya lo impide, pero un `undefined` que se colara a un
    // mensaje de error lo dejaría a medias («...asignada el  en el bloque 3»).
    expect(diaLegible(0)).toBe('día 0');
    expect(diaLegible(99)).toBe('día 99');
  });
});

describe('esChoqueDeAgenda', () => {
  it('reconoce el mensaje real del trigger del docente', () => {
    // Copiado del `raise` de `202609180001_mod3_cuadrante_aulas.sql`, con los
    // `%` ya sustituidos por lo que PostgreSQL interpola. Si alguien cambia el
    // texto en la migración, esta prueba falla y obliga a actualizar el
    // detector en vez de dejar que un 409 se degrade a 400 en silencio.
    const mensajeReal =
      'Ese docente ya tiene una clase o guardia asignada el miércoles en el bloque 5. ' +
      'Un docente no puede estar en dos sitios a la vez.';

    expect(esChoqueDeAgenda(mensajeReal)).toBe(true);
  });

  it('reconoce el mensaje real del trigger del espacio', () => {
    const mensajeReal =
      'Ese espacio ya está ocupado el lunes en el bloque 1. Dos grupos no pueden ' +
      'compartir el mismo sitio a la misma hora.';

    expect(esChoqueDeAgenda(mensajeReal)).toBe(true);
  });

  it('NO confunde un choque con un `check` corriente', () => {
    // Los dos llegan como 23514. Si esto devolviera `true`, cualquier dato mal
    // formado daría 409 y el cliente diría «esa aula está ocupada» por un
    // problema que no tiene nada que ver con la agenda.
    expect(
      esChoqueDeAgenda('El bloque no puede pasar de 12.'),
    ).toBe(false);
    expect(
      esChoqueDeAgenda('La capacidad no puede ser negativa.'),
    ).toBe(false);
  });

  it('NO confunde el trigger del período con un choque de agenda', () => {
    // `exigir_periodo_registrado` también lanza 23514 y también nombra un
    // período. Es la confusión más fácil de cometer al leer el mensaje.
    const mensajeDelPeriodo =
      'El período activo "SA26-2" no está registrado en el catálogo de lapsos. ' +
      'Créalo primero: la Regla 2 compara este valor con la sección y, si no ' +
      'corresponde a ningún lapso real, no protegería nada.';

    expect(esChoqueDeAgenda(mensajeDelPeriodo)).toBe(false);
  });

  it('no se dispara con un mensaje vacío ni con ruido', () => {
    expect(esChoqueDeAgenda('')).toBe(false);
    expect(
      esChoqueDeAgenda('duplicate key value violates unique constraint "classrooms_name_key"'),
    ).toBe(false);
  });

  it('toleraría el mensaje sin tilde, por si la colación cambia', () => {
    expect(
      esChoqueDeAgenda('Ese espacio ya esta ocupado el martes en el bloque 9.'),
    ).toBe(true);
  });
});

describe('esFechaISO', () => {
  it('acepta fechas reales', () => {
    expect(esFechaISO('2026-09-15')).toBe(true);
    expect(esFechaISO('2027-02-13')).toBe(true);
    expect(esFechaISO('2024-02-29')).toBe(true);
  });

  it('rechaza el 30 de febrero, que tiene la forma correcta y no existe', () => {
    // Es el caso que justifica la función: un patrón de expresión regular solo
    // lo aceptaría y el error llegaría a Postgres como `22007`, que el traductor
    // no reconoce y saldría como un 500 por un dato que el cliente escribió mal.
    expect(esFechaISO('2026-02-30')).toBe(false);
    expect(esFechaISO('2026-04-31')).toBe(false);
    expect(esFechaISO('2026-13-01')).toBe(false);
    expect(esFechaISO('2026-00-10')).toBe(false);
  });

  it('rechaza el 29 de febrero de un año no bisiesto', () => {
    // 2026 no es bisiesto. `new Date` lo normalizaría al 1 de marzo, y por eso
    // la comprobación compara los tres componentes y no sólo la validez.
    expect(esFechaISO('2026-02-29')).toBe(false);
  });

  it('rechaza formatos que no son AAAA-MM-DD', () => {
    expect(esFechaISO('15/09/2026')).toBe(false);
    expect(esFechaISO('2026-9-15')).toBe(false);
    expect(esFechaISO('2026-09-15T00:00:00Z')).toBe(false);
    expect(esFechaISO('')).toBe(false);
  });
});

describe('rangoDeFechasValido', () => {
  it('con las dos fechas, el fin tiene que ser posterior', () => {
    expect(rangoDeFechasValido('2026-09-21', '2027-02-13')).toBe(true);
    expect(rangoDeFechasValido('2027-02-13', '2026-09-21')).toBe(false);
  });

  it('el mismo día no es un rango válido', () => {
    // El `check` de la base exige `end_date > start_date`, no `>=`: un lapso que
    // empieza y termina el mismo día no es un lapso.
    expect(rangoDeFechasValido('2026-09-21', '2026-09-21')).toBe(false);
  });

  it('sin fechas no hay nada que ordenar', () => {
    // Las fechas nacen nulas y el centro las carga cuando las tiene. Rechazar
    // aquí un lapso sin fechas bloquearía un estado que la tabla acepta.
    expect(rangoDeFechasValido(null, null)).toBe(true);
    expect(rangoDeFechasValido(undefined, undefined)).toBe(true);
    expect(rangoDeFechasValido('2026-09-21', null)).toBe(true);
    expect(rangoDeFechasValido(null, '2027-02-13')).toBe(true);
  });
});

describe('los techos del backend y los de la base siguen de acuerdo', () => {
  it('el día llega hasta el sábado y el bloque hasta doce', () => {
    // Los dos números están duplicados en los `check` de `teacher_duties` y
    // `schedule_slots`. Esta prueba no puede leer la migración, pero deja
    // escrito de dónde salen para que el día que se cambien allí se sepa que
    // también hay que cambiarlos aquí y en `esquemas.ts`.
    expect(DIA_MAXIMO).toBe(6);
    expect(BLOQUE_MAXIMO).toBe(12);
  });
});
